Create initial power policy implementation

RobertZ2011 opened the pull request:
This PR defines the power policy service and provides an initial implementation that handles selecting the highest power sink.

felipebalbi commented on embedded-service/src/power/policy/device.rs:
```diff
@@ -0,0 +1,318 @@
+    pub async fn connect_sink(&self, capability: PowerCapability) -> Result<(), Error> {
+        info!("Device {} connecting sink", self.id.0);
+
+        let _ = self
+            .execute_device_request(RequestData::ConnectSink(capability))
+            .await?;
```
Why are you ignoring the `ResponseData`?

RobertZ2011 replied:
Done. Added a `complete_or_err` function to `ResponseData` to make it easy to convert it into a result.

felipebalbi commented on embedded-service/src/power/policy/device.rs:
```diff
@@ -0,0 +1,318 @@
+            .execute_device_request(RequestData::ConnectSink(capability))
+            .await?;
+
+        {
+            let mut lock = self.state.lock().await;
+            let state = lock.deref_mut();
+            if state.state != State::Attached {
+                warn!("Received connect sink request for device that is not attached");
+            }
+
+            state.state = State::Sink(capability);
+        }
```
The extra scope seems unnecessary. Additionally, if the current state is incorrect, should you produce an `Err` variant?

RobertZ2011 replied:
This was spillover from the implementation of the `notify_*` functions. Since those are intended to be called by the device driver I made those checks warnings on the assumption that we should just trust the driver and accept what it thinks the current state should be. But that doesn't apply to the policy-initiated request. Superseded by the type state changes.

felipebalbi commented on embedded-service/src/power/policy/device.rs:
```diff
@@ -0,0 +1,318 @@
+            .execute_device_request(RequestData::ConnectSource(capability))
+            .await?;
+
+        {
+            let mut lock = self.state.lock().await;
+            let state = lock.deref_mut();
+            if state.state != State::Attached {
+                warn!("Received connect source request for device that is not attached");
+            }
+
+            state.state = State::Source(capability);
+        }
```
ditto.

felipebalbi commented on embedded-service/src/power/policy/device.rs:
```diff
@@ -0,0 +1,318 @@
+
+        let _ = self.execute_device_request(RequestData::Disconnect).await?;
+
+        {
+            let mut lock = self.state.lock().await;
+            let state = lock.deref_mut();
+
+            if !matches!(state.state, State::Sink(_)) && !matches!(state.state, State::Source(_)) {
+                warn!(
+                    "Disconnect request for device that is not sourcing or sinking, state: {:?}",
+                    state.state
+                );
+            }
+
+            state.state = State::Attached;
+        }
```
ditto, and because there's so much duplication, it seems like you could extract into a helper that takes a closure argument. Something along the lines of:

```rust
    fn validate<F>(self, state: State, f: F) -> T
    where
        F: FnOnce(State) -> T,
    {
        f(state)
    }

    fn foo() {
        let mut lock = self.state.lock().await;
        let state = lock.deref_mut();

        state.state = self.validate(state, |state| {
            if !matches!(state.state, State::Sink(_)) && !matches!(state.state, State::Source(_)) {
                warn!(
                    "Disconnect request for device that is not sourcing or sinking, state: {:?}",
                    state.state
                );
                Err(Error::InvalidStateTransition(state.state))
            } else {
                Ok(State::Attached);
            }
        })?;
    }
```

felipebalbi commented on embedded-service/src/power/policy/device.rs:
```diff
@@ -0,0 +1,318 @@
+/// Struct that provides functions to send requests to the power policy from a specific device
+pub struct PolicyInterface<'a>(&'a Device);
+
+impl PolicyInterface<'_> {
```
I see an option to encode the state as typestate, then compiler can prove valid state transitions for you. That being said, it might be difficult to work with resulting API, because you will need to consume `self` in order to change state.

RobertZ2011 replied:
I created two wrapper types: `state_machine::Device` and `state_machine::Policy` that use type states to control the actions and transitions available both to the power policy and individual devices. Wrappers are produced by the `try_*_state_machine` functions which return results to allow the caller to decide how to handle encountering an unexpected state. This should make working with each side less error prone without needing to transform the type of the overall device.

jerrysxie commented on power-policy/src/lib.rs:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    /// Determines and connects the best sink
+    async fn select_sink(&self) -> Result<(), Error> {
```
Would it be possible to refactor this function a bit to make it more clear? Maybe break it up into sub-functions? It is very tough to read.

Let me see if I grasp what it is trying to do:
1. Searching through existing devices to find the best sink if it is the disconnecting sink ignore
2. If the best sink is same and it is still available, then okay. Else disconnect the sink?
3. Connect the new best sink?

The second part doesn't make sense to me. We are not considering the sink being disconnected in the first search. Then it wouldn't be possible for the best sink to be the device being disconnected, no?

Wouldn't the logic be simpler to record the current best sink, if it is being removed, then find the best sink in the remaining sink, disconnect the current sink and connect the new best sink. Else just keep the best sink as it is.

jerrysxie replied:
It seems like the logic in this function is this way, because it is trying to handle lots of different cases. Can we simplify this a bit? It might be more verbose to handle different case separately, but it might be more readable.

RobertZ2011 replied:
I split the function into two subfunctions and did some unnesting of braces to clarify things. Disconnecting the current sink is done because it may still be present but we're switching to a higher power charger.

jerrysxie commented on power-policy/src/lib.rs:
```diff
@@ -0,0 +1,256 @@
+                    .context
+                    .get_device(current_sink.device_id)
+                    .await?
+                    .try_policy_state_machine::<sm::Sink>()
```
Why would this state transition fail if it is the current sink?

RobertZ2011 replied:
If the current sink has been unplugged then we won't need to disconnect it.

RobertZ2011 commented on embedded-service/src/power/policy/policy.rs:
```diff
@@ -0,0 +1,169 @@
+    /// Returns an InvalidResponse error if the response is not complete
+    pub fn complete_or_err(self) -> Result<(), Error> {
+        match self {
+            ResponseData::Complete => Ok(()),
```
Yes if more values are added to `ResponseData`. No policy requests return data so this is currently the only value in the enum. I did try marking the enum `#[non_exhaustive]` and adding a wildcard match arm, but I was still getting warnings about the wildcard being unreachable so I decided to leave it as is for now.

RobertZ2011 commented on embedded-service/src/power/policy/device.rs:
```diff
@@ -0,0 +1,424 @@
+    pub struct Policy<'a, S: Kind> {
+        device: &'a super::Device,
+        _state: core::marker::PhantomData<S>,
+    }
```
I spent some time trying out this approach, but I think the current method is better. We'd need at least three power role values (NoRole, Source, Sink) and three power state values (Detached (not present), Idle (present but not in use), Connected (present and in use)). But this allows for states that are valid to the compiler, but not to the application (e.g. (Detached, Source) and (Detached, Sink)).  I also don't think the (Idle, Source) and (Idle, Sink) states make sense because the device doesn't really have a role until the power policy decides what it wants. It would be possible to use those states to indicate a device that can only act as a source and a device that can only act as a sink, but we'd also need a way to indicate dual-role devices and things start getting really complicated.

RobertZ2011 commented:
Also updated the documentation to explain the design of the service better and point out the core structs there. I also did a rename. The source and sink nomenclature was borrowed from the PD standard and I realized that it can be easy to confuse things. Renamed source->provider, sink->consumer and attached->idle.

felipebalbi commented on embedded-service/src/power/policy/action/device.rs:
```diff
@@ -0,0 +1,97 @@
+    /// Detach the device
+    pub async fn detach(self) -> Result<Device<'a, Detached>, Error> {
+        info!("Received detach from device {}", self.device.id().0);
+        self.device.set_state(device::State::Detached).await;
```
This line implies that you're tracking state via typestate and via a dedicated field within `Device`. Why the duplication?

RobertZ2011 replied:
This is because we can't change the overall type of `Device` since it needs to live in something like a `OnceLock` or `StaticCell` and because the power policy implementation needs a fixed type to work with. So these action types provide checking on a local level and the field keeps track of the state on a global level without having to put the type state on `Device` itself.

felipebalbi commented on embedded-service/src/power/policy/action/policy.rs:
```diff
@@ -0,0 +1,78 @@
+    }
+}
+
+impl<'a> Policy<'a, ConnectedConsumer> {
+    /// Disconnect this device
+    pub async fn disconnect(self) -> Result<Policy<'a, Idle>, Error> {
+        self.disconnect_internal().await?;
+        Ok(Policy::new(self.device))
+    }
+}
+
+impl<'a> Policy<'a, ConnectedProvider> {
+    /// Disconnect this device
+    pub async fn disconnect(self) -> Result<Policy<'a, Idle>, Error> {
+        self.disconnect_internal().await?;
+        Ok(Policy::new(self.device))
+    }
+}
```
Duplicated code here. Maybe add a small trait that both `ConnectedConsumer` and `ConnectedProvider` implement?

RobertZ2011 replied:
I tried this, but the traits end up a bit more complicated than I'd like. Given that the duplication is small I'm going to leave it as is.

RobertZ2011 commented on power-policy-service/src/lib.rs:
```diff
@@ -0,0 +1,264 @@
+    }
+
+    async fn process_notify_attach(&self) -> Result<(), Error> {
+        self.context.send_response(Ok(policy::ResponseData::Complete)).await;
```
Yeah, the provider logic will fill out the other empty handler. This one will remain just because there's not anything to do aside from acknowledging the notification.

RobertZ2011 commented on power-policy-service/src/lib.rs:
```diff
@@ -0,0 +1,264 @@
+            }
+
+            self.comms_notify(CommsMessage {
+                data: CommsData::ConsumerConnected(current_consumer.device_id, new_consumer.power_capability),
```
Done, good catch!

RobertZ2011 commented on embedded-service/src/power/policy/action/device.rs:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    /// Detach the device
+    pub async fn detach(self) -> Result<Device<'a, Detached>, Error> {
+        info!("Received detach from device {}", self.device.id().0);
+        self.device.set_state(device::State::Detached).await;
+        self.device.update_consumer_capability(None).await;
+        policy::send_request(self.device.id(), policy::RequestData::NotifyDetached)
+            .await?
+            .complete_or_err()?;
+        Ok(Device::new(self.device))
+    }
```
This transition is valid. It simplifies the device recover logic a bit to always have the `detach` action available. I updated the docs to include this transition.

RobertZ2011 commented on embedded-service/src/power/policy/action/device.rs:
```diff
@@ -0,0 +1,97 @@
+        .await?
+        .complete_or_err()
+    }
+}
```
The transition from `Idle` is driven from the policy side not the device side. The corresponding policy `Idle` action contains those functions.

RobertZ2011 commented:
Fixed a few small bugs and added types to contain any action after my experiences with the PD controller integration.

RobertZ2011 commented on power-policy-service/src/lib.rs:
```diff
@@ -0,0 +1,266 @@
+    async fn find_highest_power_consumer(&self) -> Result<Option<ConsumerState>, Error> {
+        let mut best_consumer = None;
+
+        for node in self.context.devices().await {
+            let device = node.data::<Device>().ok_or(Error::InvalidDevice)?;
+
+            // Update the best available consumer
+            best_consumer = match (best_consumer, device.consumer_capability().await) {
+                // Nothing available
+                (None, None) => None,
+                // No existing consumer
+                (None, Some(power_capability)) => Some(ConsumerState {
+                    device_id: device.id(),
+                    power_capability,
+                }),
+                // Existing consumer, no new consumer
+                (Some(_), None) => best_consumer,
+                // Existing consumer, new available consumer
+                (Some(best), Some(available)) => {
+                    if available > best.power_capability {
+                        Some(ConsumerState {
+                            device_id: device.id(),
+                            power_capability: available,
+                        })
+                    } else {
+                        best_consumer
+                    }
+                }
+            };
+        }
```
Definitely a fold, but I don't think it's worth it in this case. We'd need space to store a collection of the available power numbers but the amount of devices isn't fixed and I'd like to avoid limiting it with a constant in order to use something like a heapless vec.
